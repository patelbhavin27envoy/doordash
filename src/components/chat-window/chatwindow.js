import React, {Component } from 'react';
import { withRouter } from 'react-router-dom';

import './chatwindow.scss';
import { Groupchat } from "../group-chat";
import { Rooms } from "../rooms";
import {apiService} from "../../services/apiService";


/* For Redux
* import { connect } from 'react-redux'
* import { actions } from '../../actions'
*/

class ChatWindow extends Component {

    constructor(props) {

        super(props);
        let uName = null;
        if(this.props.location && this.props.location.state) {
            uName = this.props.location.state.userName;
        }
        this.state = {
            rooms: '',
            room: {id:1, name:'Analytics'},
            userName: uName,
            error: null,
            users: this.props.users
        };

        this.roomChangeHandler = this.roomChangeHandler.bind(this);
        this.listOfUsers = this.listOfUsers.bind(this);
    }

    componentDidMount() {

        this.setState({
            'error': null,
            'loading' : true
        });

        this.roomChangeHandler(this.state.room);
    }

    isLoading() {
        return this.state.error === null && this.state.venues === null;
    }

    listOfUsers (users) {
        let roomUsers = '';
        users.forEach((user,index) => {
            if(index != users.length-1) {
                roomUsers += user + (', ');
            } else {
                roomUsers += user ;
            }
        });
        return roomUsers;
    }

    roomChangeHandler(room) {

        this.setState({
            room : room
        });

        apiService()
            .getRoomDetails(room.id)
            .then((data) => {

                this.setState({
                    'users': data.users,
                    'error': null,
                    'loading' : false
                });

            })
            .catch((error) => {

                this.setState({
                    rooms : null,
                    error: 'Sorry! we had some problem with your request.',
                    'loading' : false
                })
            });
    }

    render(){

        const {rooms, userName, error, users, room} = this.state;
        let roomUsers  = '...';
        if(users) {
            roomUsers = this.listOfUsers(users);
        }

        if(roomUsers) {
            return (
                <div className="row container-flex flex-row">

                    <Rooms userName={userName} roomChange={ (room) => {this.roomChangeHandler(room)} } />

                    <div className="right">
                        <div className="chat-header">
                            <div className="room-window">
                               <Groupchat userName={userName} room={this.state.room} roomUsers={roomUsers} />
                            </div>
                        </div>
                    </div>

                </div>
            )

        }
    }

}

export default  withRouter(ChatWindow);
