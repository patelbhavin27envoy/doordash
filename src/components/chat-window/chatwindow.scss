@import '../../index.scss';

.right {
  flex: 1;
}
