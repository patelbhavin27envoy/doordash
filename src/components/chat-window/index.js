import ChatWindow from './chatwindow'

export {
    ChatWindow
}
