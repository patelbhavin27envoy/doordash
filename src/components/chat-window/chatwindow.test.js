import React from 'react';
import renderer from 'react-test-renderer';

import { shallow, mount } from "enzyme";
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

import { ChatWindow } from '../chat-window'
import { MemoryRouter } from 'react-router-dom';

describe("ChatWindow Component :: ", () => {

    let userName = "test-1";
    let roomChangeHandler = (room) => {
        return room;
    };

    it("should render", () => {

        // Act
        let component = renderer.create(
            <MemoryRouter>
                <ChatWindow />
            </MemoryRouter>
        ).toJSON();
        // snapshot
        expect(component).toMatchSnapshot();


        // Act
        component = mount(
            <MemoryRouter>
                <ChatWindow />
            </MemoryRouter>
        );

        expect(component.find('.error').text())
            .toContain('Sorry ! something went wrong');
        expect(component.find('.chatList').text())
            .toContain("");
    });


    it("should make the service call for rooms", () => {

        // define the response
        const usersData = [ 'user-1', 'user-2', 'user-3'];

        // let's mock the service call
        const endPoint = 'http://localhost:8080/api/rooms';
        const mockJsonPromise = Promise.resolve(usersData);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        jest.spyOn(global, 'fetch').mockImplementation(() => { return mockFetchPromise }); // 4
        let component = mount(
            <MemoryRouter>
                <ChatWindow  users={usersData} />
            </MemoryRouter>
        );

        component.update();
        expect(component.find('.error').text())
            .toContain('Sorry ! something went wrong');
        expect(component.find('.chatList').text())
            .toContain("");

        const rooms = [
            {"name":"Analytics","id":1},
            {"name":"Business","id":2},
            {"name":"Design","id":3},
            {"name":"Engineering","id":4},
            {"name":"HR","id":5},
            {"name":"Operations","id":6},
            {"name":"Special Ops","id":7}];

        component.setState({
            users: usersData ,
            room: {
                id: 1, name : 'Analytics'
            },
            rooms: rooms
            });
        component.update();

        console.log(component.debug());
        expect(component.find('.users').text())
            .toContain("user-1, user-2, user-3");
    });



});

