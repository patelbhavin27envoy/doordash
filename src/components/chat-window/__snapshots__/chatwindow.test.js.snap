// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatWindow Component ::  should render 1`] = `
<div
  className="row container-flex flex-row"
>
  <div
    className="left"
  >
    <div
      className="wrapper"
    >
      <div
        className="sidebar-header"
      >
        <div
          className="user"
        >
          <h3
            className="name"
          />
          <div
            className="online-time"
          >
            was online 1 hour before
          </div>
        </div>
      </div>
      <ul
        className="list-unstyled components justify-content-center"
      >
        <div
          className="rooms"
        >
          <div
            className="error"
          >
             Sorry ! something went wrong 
          </div>
        </div>
      </ul>
    </div>
  </div>
  <div
    className="right"
  >
    <div
      className="chat-header"
    >
      <div
        className="room-window"
      >
        <div
          className="chatWindow"
        >
          <div
            className="header"
          >
            <h3
              className="name"
            >
              Analytics
            </h3>
            <div
              className="users"
            >
              ...
            </div>
          </div>
          <ul
            className="chat chatList"
          />
          <div
            className="chatInputWrapper"
          >
            <form
              onSubmit={[Function]}
            >
              <input
                className="textarea input"
                name="usermessage"
                onChange={[Function]}
                placeholder="Enter your message..."
                type="text"
              />
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
